The wiki holds documentation most relevant for develop. For information on a specific version of Truvari, see [`docs/`](https://github.com/spiralgenetics/truvari/tree/develop/docs)

Citation:  
English, A.C., Menon, V.K., Gibbs, R.A. et al. Truvari: refined structural variant comparison preserves allelic diversity. Genome Biol 23, 271 (2022). https://doi.org/10.1186/s13059-022-02840-6

# Index

- [[Updates|Updates]]
- [[Installation|Installation]]
- [[Development|Development]]

Truvari Commands:  
- Benchmarking  
  - [[bench|bench]] - Performance metrics from comparison of two VCFs  
  - [[refine|refine]] - Automated bench result refinement with phab  
- Merging  
  - [[collapse|collapse]]     Collapse redundant VCF entries  
  - [[phab|phab]]         Variant harmonization using MSA  
- Analysis  
  - [[consistency|consistency]]  Consistency report between multiple VCFs  
  - [[stratify|stratify]]     Count VCF entries inside BED regions  
  - [[vcf2df|vcf2df]]       Turn VCF into pandas DataFrame  
  - [[stratp|Stratp-Test]]       Stratification performance test  
- Annotation  
  - [[anno|anno]]         VCF Annotations  
- Misc  
  - [[segment|segment]]      Normalization of SVs into disjointed genomic regions  
  - [[divide|divide]]       Divide a VCF into independent shards  
  - [[ga4gh|ga4gh]]        Convert Truvari result to GA4GH  

